examples/vision/image_enhancer.py

```diff
@@ -0,0 +1,276 @@
+"""
+Title: Document Images Restoration/Enhancer.
+Author: [Anish B](https://twitter.com/anishhacko)
+Date created: 2021/01/18
+Last modified: 2022/01/21
+Description: Example of Document Image Restoration coupled with Tesseract OCR engine to improve text readability.
+"""
+"""
+## Introduction
+
+This example explains a "Image Enhancer/Denoiser" using UNet architecture. The problem we
+focus here is enhancing the document images which are deteriorated through external
+degradations like blur, corrupt text blocks etc... ,we have followed up on few seminal
+papers as reference and used simple pretrained UNet as encoder and Efficient Sub-Pixel
+CNN as decoder. we do not attempt super-resolution as primary task hence we return output
+image in the same dimension as input. we compute WER (Word Error Rate) metric to evaluate
+model performance helping us to assess the text information gain in real world scenarios
+when coupled with tesseract OCR engine.
+
+**References:**
+- [Enhancing OCR Accuracy with Super
+Resolution](https://cdn.iiit.ac.in/cdn/cvit.iiit.ac.in/images/ConferencePapers/2018/ocr_Ankit_Lat_ICPR_2018.pdf)
+Resolution](https://cdn.iiit.ac.in/cdn/cvit.iiit.ac.in/images/ConferencePapers/2018/ocr_Ankit_Lat_ICPR_2018.pdf)
+- [Improving the Perceptual Quality of Document
+Images Using Deep Neural
+Network](http://mile.ee.iisc.ac.in/publications/softCopy/DocumentAnalysis/ISNN_11page_65.pdf)
+Network](http://mile.ee.iisc.ac.in/publications/softCopy/DocumentAnalysis/ISNN_11page_65.pdf)
+"""
+
+"""
+## Additional Set-up
+"""
+
+# !pip install pybind11
+# !pip install fastwer
+# !pip install pytesseract
+# !sudo apt install tesseract-ocr
+
+import fastwer
+import pytesseract
+import numpy as np
+from glob import glob
+import tensorflow as tf
+from tensorflow import keras
+import matplotlib.pyplot as plt
+from tensorflow.keras import callbacks
+from tensorflow.keras.applications import resnet_v2
+
+"""
+## Dataset
+we have created a synthetic document dataset that depicts several real world document
+noises, we have used blurs and morphological filters to reconstruct real world noises.
+"""
+
+"""shell
+!gdown https://drive.google.com/uc?id=1jQfHJUUQcpktQcjIzitvb4SldKwqcAMr
+!unzip -q data.zip
+"""
+
+"""
+## Dataset Preprocessing
+"""
+
+train_source = sorted(glob("data/source/*.jpg"))[:-80]
+train_targets = sorted(glob("data/target/*.jpg"))[:-80]
+
+val_source = sorted(glob("data/source/*.jpg"))[-80:]
+val_targets = sorted(glob("data/target/*.jpg"))[-80:]
+
+height, width = 352, 608
+batch_size = 4
+
+
+def resize_with_pad(image, height, width):
+    return tf.image.resize_with_pad(
+        image=image, target_height=height, target_width=width
+    )
+
+
+def data_preprocess(source, target):
```

Comment:
@anish9 
In this function, you use, for example, `tf.io.deconde_png` twice, but as the source and target are going through the same operation, I think all of that operation can be placed into a single function.  For example: 

```python
def pre_process(input):
     x = tf.io.read_file(input...)
     x = tf.io.decode_png ...
     x = tf.image.resize_with_pad ...
     return x 

source_image = pre_process( source )
target_image = pre_process( target )
````